### ISSUE TITLE
“libJolt.a” uses custom architectures. Would you like to update to build with standard architectures?

### ISSUE THREAD

**Issue by PinkFatMan** [jrouwe/JoltPhysics#619]
Using the master branch

After generating an Xcode solution on the Mac. Open Build prompt

"LibJolt. a" uses custom architectures Would you like to update to build with standard architectures?

The current configuration builds for the Intel architecture, and does not include the Apple Silicon architecture Choose "Build" to keep the current configuration

Selecting Build or Update and Build will result in the following error

Clang: error: argument unused during compilation: '- mavx2' [- Error, - Wunused command line argument]

Clang: error: argument unused during compilation: '- mbmi' [- Werr, - Wunused command line argument]

Clang: error: argument unused during compilation: '- mapcnt' [- Werr, - Wunused command line argument]

Class: error: argument unused during compilation: '- mlzcnt' [- Error, - Wunused command line argument]

Clang: error: argument unused during compilation: '- mf16c' [- Werr, - Wunused command line argument]

Clang: error: argument unused during compilation: '- mfma' [- Werr, - Wunused command line argument]

**Comment by jrouwe** [comment]
Hello,

Do you have an intel or an Apple silicon Mac? 

I'm guessing intel because I tested on an M1 and never on an intel Mac. I don't own a Mac, so I need to rent a virtual machine to test things, so I'm hoping you can help out.

The errors indicate that the architecture is not correctly determined by CMake and I've written a tentative fix here:

https://github.com/jrouwe/JoltPhysics/tree/bugfix/xcode

Can you try it out?

**Comment by PinkFatMan** [comment]
Nice to help you test it

I tried using the bugfix/xcode branch and encountered many errors similar to those in my screenshot
<img width="1155" alt="image" src="https://github.com/jrouwe/JoltPhysics/assets/121303344/e9dd0a58-70d2-4b1f-a529-b0f94f5bce9d">

**Comment by jrouwe** [comment]
Are you using an intel or an Apple Silicon Mac?
What version of XCode are you using?
Which clang version is that using?

**Comment by mihe** [comment]
I happen to have an Intel-based Mac, and I can confirm that generating an Xcode project and building from Xcode (on latest macOS/Xcode) works just fine when using either of the two `cmake_xcode_*.sh` scripts on the `master` branch. You do however end up targeting x86 by default, since you need to explicitly pass `-D"CMAKE_OSX_ARCHITECTURES=x86_64;arm64"` to build universal binaries, as noted in #315.

> "LibJolt.a" uses custom architectures. Would you like to update to build with standard architectures?

Judging by this message and the subsequent errors I'm assuming @PinkFatMan clicked yes on this prompt, which (without never having seen the prompt myself) presumably makes Xcode switch over to instead target a universal binary, in which case all the x86-specific flags generated by CMake will indeed output a bunch of warnings which then turn into errors, similar to #315.

Seeing as how targeting only Intel-based Apple devices is becoming largely irrelevant, it might be pragmatic to just have `-D"CMAKE_OSX_ARCHITECTURES=x86_64;arm64"` be added to both `cmake_xcode_*.sh` scripts. Otherwise, when building on an Intel-based Mac, you'll by default only target x86, which doesn't make a whole lot of sense these days, and is presumably what triggered the prompt mentioned in this issue. This does mean that the resulting binaries would end up being a bit bigger, but I assume most/all real-world applications build universal binaries anyway.

**Comment by jrouwe** [comment]
Thanks for testing @mihe. Are you sure that option makes sense for iOS too? Unless you can run in emulator and on device from 1 package, I wouldn't expect x86 and arm to be a combo there.

I've added the option for macOS to: https://github.com/jrouwe/JoltPhysics/tree/bugfix/xcode. Does this work?

**Comment by mihe** [comment]
> I wouldn't expect x86 and arm to be a combo there.

Yeah, no, you're right. For some reason I was dead sure there were Intel-based iPads floating around, but that seems to never have been the case. My bad.

**Comment by PinkFatMan** [comment]
Thank you to @mihe  and @jrouwe  for their help. I use the latest bugfix/xcode branch and upgrade my system and xcode versions. After executing the sh command, I can generate a solution and compile it successfully. But the compiled libJolt. a file is a large 170MB, even if I modify Generate Debug Symbols to No, it still has 48MB. May I ask which part of my setting is not correct?

**Comment by jrouwe** [comment]
I have no idea. I just compiled without symbols on Linux with clang and libJolt.a is only 8 MB (and the UnitTests executable only 4 MB). I'm not really sure if this is 'incorrect' as my Windows build produces a Jolt.lib of 101 MB but the UnitTests.exe is only 2 MB.

**Comment by mihe** [comment]
> I'm not really sure if this is 'incorrect' as my Windows build produces a Jolt.lib of 101 MB but the UnitTests.exe is only 2 MB.

You get similar numbers on macOS as well, even with universal binaries and debug symbols enabled. Since static libraries are just bundles of object files, that can presumably contain any arbitrary compiler metadata, I don't think optimizing them for size is going to be very meaningful.

**Comment by PinkFatMan** [comment]
Thank you for your reply. It is true that the compiled Static library file will be large. If converted to a dynamic library, it will be very small. It should be a normal phenomenon

**Comment by jrouwe** [comment]
I think this issue is resolved now, if not let me know.
